src/Gui/Tree.cpp

```diff
@@ -1033,6 +1054,39 @@ void TreeWidget::onFinishEditing()
     }
 }
 
+// add childs of the selected tree object to selection
+void TreeWidget::onSelectAllChildren()
+{
+    // we only have this context menu entry if the selection is within one document
+    // so we can get the active document once
+    auto doc = Gui::Application::Instance->activeDocument()->getDocument();
+    auto docName = doc->getName();
+    // if the selected object is a document
+    if (this->contextItem && this->contextItem->type() == DocumentType) {
+        DocumentItem* docitem = static_cast<DocumentItem*>(this->contextItem);
+        //App::Document* doc = docitem->document()->getDocument();
+        std::vector<App::DocumentObject*> obj = doc->getObjects();
+        for (std::vector<App::DocumentObject*>::iterator it = obj.begin(); it != obj.end(); ++it)
+            Selection().addSelection(docName, (*it)->getNameInDocument());
+    }
+    // it can be an object
+    else {
+        QList<QTreeWidgetItem*> items = this->selectedItems();
+        for (QList<QTreeWidgetItem*>::iterator it = items.begin(); it != items.end(); ++it) {
+            if ((*it)->type() == ObjectType) {
+                DocumentObjectItem* objitem = static_cast<DocumentObjectItem*>(*it);
+                App::DocumentObject* obj = objitem->object()->getObject();
+                auto subObjectList = obj->getOutList();

```

Comment:
getOutList() returns the list of objects a feature depends on but in general these are not sub-items in the tree view. The view provider offers the method claimChildren() that gives the list of sub-objects in the tree view.

But even this won't produce the expected behaviour in many cases because an object can be claimed by several view providers as child and thus it appears several times in the tree view.

Example:
Create a box and cylinder in Part. Now select them and create a fusion, a cut and an intersection. Now select one of the boolean operation objects and invoke your function. In most cases it selects the sub-object of two different boolean operations.

Thus, this function should not use the document & view providers at all but only use functions of QTreeWidgetItem.

Reply:
> getOutList() returns the list of objects a feature depends on but in general these are not sub-items in the tree view. 

But this is what the feature should do. The children depending on the selected object should be selected. That does not necessarily mean that all objects are selected that are subobjects in the tree.

Maybe the bug reporter intended to select it differently. But selfish as I am ;-) I was focused on my needs and this is to select the depending objects. Your example is a good example:

> Create a box and cylinder in Part. Now select them and create a fusion, a cut and an intersection. Now select one of the boolean operation objects and invoke your function. In most cases it selects the sub-object of two different boolean operations.

I made a fusion and subsequently a cut and when selecting the children of the cut, I get this:
![Kt3nShrLzr](https://user-images.githubusercontent.com/1828501/102155488-ecd1bd00-3e7b-11eb-8dea-458259d7b5f7.gif)

This is correct because I did the selection to delete the cut. So after the deletion of the cut, the fusion and thus its childs must remain. Therefore not all subobjects in the tree should be selected. I even don't see a good use case why one wants to select all tree subobjetcs.

Reply:
Depending on what decision you and @wwmayer come to in this regard, it might make sense to change the text to "Select all dependents" instead. I interpreted "children" in the GUI tree view sense.

Reply:
>  I forgot that the childs can have childs as well. I'll try to fix this.

I fixed this now. When I select the children of the cut and then press DEL the cut an all its children are correctly deleted, no item remains.

There is however a general display issue:
![61rmNHaxr3](https://user-images.githubusercontent.com/1828501/102162076-d120e380-3e88-11eb-9284-d38ca6bb65be.gif)

So no matter if I use my new selection feature or select all items manually, items that appear twice in the tree will only be highlighted once.

Reply:
> Depending on what decision you and @wwmayer come to in this regard, it might make sense to change the text to "Select all dependents" instead. I interpreted "children" in the GUI tree view sense.

+1

I find the term "children" quite confusing in this context when the function is about to select the dependents. A dependent is not necessarily a child item in the tree view but can be a sibling item. So, all the function and variable names should make this clear and thus avoid the term child or children and use dependents instead.

Reply:
> I find the term "children" quite confusing in this context when the function is about to select the dependents...

Indeed. I changed this now. Using the term "dependents" will make it hopefully more clear what it is about. For example here, I selected all dependents of the ProjGroup to quickly see what it depends on:

![FreeCAD_xE0SkOi3ZG](https://user-images.githubusercontent.com/1828501/102726052-cd271280-431b-11eb-925d-42b40d4663ed.png)